parser/src/err.rs

```diff
@@ -0,0 +1,121 @@
+use crate::ast::Expr;
+use crate::parser::ParseResult;
+use context::source::Location;
+use lexer::token::Token;
+
+/// An error that occurs during parsing.
+#[derive(Debug, PartialEq)]
+pub struct ParseError {
+    pub message: String,
+    pub position: Location,
+    pub kind: ParseErrorKind,
+}
+
+/// A builder to position an error that covers multiple tokens.
+#[derive(Debug, PartialEq)]
+pub(crate) struct ErrorContext<'a> {
+    pub(crate) from: Token<'a>,
+    pub(crate) to: Token<'a>,
+}
+
+impl<'a> From<Token<'a>> for ErrorContext<'a> {
+    fn from(token: Token<'a>) -> Self {
+        Self {
+            from: token.clone(),
+            to: token,
+        }
+    }
+}
+
+impl<'a> From<std::ops::Range<Token<'a>>> for ErrorContext<'a> {
+    fn from(range: std::ops::Range<Token<'a>>) -> Self {
+        Self {
+            from: range.start,
+            to: range.end,
+        }
+    }
+}
+
+/// The kind of error that occurred.
+///
+/// This is used to categorize the error and to provide more information
+/// when available.
+#[derive(Debug, PartialEq, Clone)]
+pub enum ParseErrorKind {
+    /// A specific token was expected, but another token was found.
+    Excepted(&'static str),
+
+    /// A unexpected token was found in the current context.
+    Unexpected,
+
+    /// A group has been opened, but not closed.
+    ///
+    /// This reports the location of the opening token.
+    Unpaired(Location),
+
+    /// A token cannot be parsed.
+    NotParsable,
```

Comment:
I think that`NotParsable` is a bit ambiguous in a context of a language parser. Maybe try refactoring with `NumberFormat`, `NumberNotParsable` would be more appropriate as i saw that you used this to signify that a number literal could not be parsed.